### ISSUE TITLE
nxos_ospf_interfaces module does not activate OSPF on L3 physical interface

### ISSUE THREAD

**Issue by ChristopherJHart** [ansible-collections/cisco.nxos#205]
##### SUMMARY
The nxos_ospf_interfaces module doesn't appear to activate an OSPF process on an L3 physical interface.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
nxos_ospf_interfaces

##### ANSIBLE VERSION

```paste below
chart2@chart2-ubuntu-playground:~/Documents/Ansible/ansible-ospf-bug$ ansible --version
715893 1606928906.14279: starting run
ansible 2.10.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/chart2/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/chart2/.local/lib/python3.8/site-packages/ansible
  executable location = /home/chart2/.local/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]

chart2@chart2-ubuntu-playground:~/Documents/Ansible/ansible-ospf-bug$ ansible-galaxy collection list
715907 1606928962.57991: starting run
715907 1606928962.91980: Validate TLS certificates for https://galaxy.ansible.com: True

# /home/chart2/.ansible/collections/ansible_collections
Collection        Version
----------------- -------
ansible.netcommon 1.4.1
cisco.nxos        1.3.1
```

##### CONFIGURATION

```paste below
chart2@chart2-ubuntu-playground:~/Documents/Ansible/ansible-ospf-bug$ ansible-config dump --only-changed
715936 1606928977.32609: starting run
DEFAULT_DEBUG(env: ANSIBLE_DEBUG) = True
DEFAULT_FORKS(/etc/ansible/ansible.cfg) = 50
DEFAULT_GATHER_TIMEOUT(/etc/ansible/ansible.cfg) = 60
DEFAULT_LOG_PATH(env: ANSIBLE_LOG_PATH) = /home/chart2/Documents/Ansible/ansible-ospf-bug/ansible.log
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
PERSISTENT_CONNECT_TIMEOUT(/etc/ansible/ansible.cfg) = 180
```

##### OS / ENVIRONMENT
Issue reproduced against a Nexus 92160YC-X running NX-OS software release 9.3(6).

##### STEPS TO REPRODUCE
Consider the below playbook:
```yaml
- name: Configure OSPF
  connection: network_cli
  gather_facts: false
  hosts: all
  tasks:
    - name: Configure OSPF on underlay interface
      cisco.nxos.nxos_ospf_interfaces:
        config:
        - name: "Ethernet1/1"
          address_family:
          - afi: ipv4
            processes:
            - process_id: UNDERLAY
              area:
                area_id: 0.0.0.0
```
The Nexus has an OSPF process called "UNDERLAY" already configured:
```
N9K# show running-config ospf 

!Command: show running-config ospf
!Running configuration last done at: Wed Dec  2 17:06:15 2020
!Time: Wed Dec  2 17:10:24 2020

version 9.3(6) Bios:version 07.68 
feature ospf

router ospf UNDERLAY
```
The current configuration of Ethernet1/1 is as follows:
```
N9K# show running-config interface Ethernet1/1

!Command: show running-config interface Ethernet1/1
!Running configuration last done at: Wed Dec  2 17:06:15 2020
!Time: Wed Dec  2 17:11:23 2020

version 9.3(6) Bios:version 07.68 

interface Ethernet1/1
  ip address 192.168.1.1/24
  no shutdown
```
I would expect that the aforementioned playbook would activate the UNDERLAY OSPF process on Ethernet1/1. Interestingly, the playbook returns "OK" instead of "Changed":
```
chart2@chart2-ubuntu-playground:~/Documents/Ansible/ansible-ospf-bug$ ansible-playbook -i inventories/hosts playbook.yml

PLAY [Configure OSPF] *****************************************************************************************************************************************************

TASK [Configure OSPF on underlay interface] *******************************************************************************************************************************
ok: [N9K]

PLAY RECAP ****************************************************************************************************************************************************************
N9K                        : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
After execution, the configuration of Ethernet1/1 has not been modified:
```
N9K# show running-config interface Ethernet1/1

!Command: show running-config interface Ethernet1/1
!Running configuration last done at: Wed Dec  2 17:06:15 2020
!Time: Wed Dec  2 17:12:26 2020

version 9.3(6) Bios:version 07.68 

interface Ethernet1/1
  ip address 192.168.1.1/24
  no shutdown
```
A playbook that fully configured a Nexus with the above configuration can be found here (this also reproduces the issue, as running this playbook multiple times does not activate OSPF on Ethernet1/1):
```yaml
- name: Configure OSPF
  connection: network_cli
  gather_facts: false
  hosts: all
  tasks:
    - name: Configure OSPF underlay process
      cisco.nxos.nxos_ospfv2:
        config:
          processes:
          - process_id: UNDERLAY 
    - name: Configure underlay interface
      cisco.nxos.nxos_interfaces:
        config:
        - name: "Ethernet1/1"
          mode: layer3
          enabled: true
    - name: Configure IP on underlay interface
      cisco.nxos.nxos_l3_interfaces:
        config:
        - name: "Ethernet1/1"
          ipv4:
          - address: "192.168.1.1/24"
    - name: Configure OSPF on underlay interface
      cisco.nxos.nxos_ospf_interfaces:
        config:
        - name: "Ethernet1/1"
          address_family:
          - afi: ipv4
            processes:
            - process_id: UNDERLAY
              area:
                area_id: 0.0.0.0
```

##### EXPECTED RESULTS
I would expect that the UNDERLAY OSPF process is activated under the Ethernet1/1 interface after this playbook is executed.

##### ACTUAL RESULTS
The OSPF process is not activated under Ethernet1/1.

Here is a gist with verbose output with ANSIBLE_DEBUG set: https://gist.github.com/ChristopherJHart/27422e2374263f626c4442e171976cfe

Here is a gist with the contents of the Ansible logfile: https://gist.github.com/ChristopherJHart/606fa0bbe5a7b6f198428e51b7ea18ee

**Comment by NilashishC** [comment]
@ChristopherJHart Thanks for reporting this issue and sharing the debug logs. Both the plays that you mentioned worked fine for me on a NX-OS 9.3.3. Since this is a common usage, we have test cases to cover this and none of those are failing right now. 

The `ospf_interfaces` module works on the output of `show running-config | section '^interface'`. Do you mind sharing that from the target device so that I can debug this further? The persistent log messages from the playbook run would also help. The logs you shared don't have that option enabled. Details on that can be found [here](https://docs.ansible.com/ansible/latest/network/user_guide/network_debug_troubleshooting.html#enabling-networking-device-interaction-logging). After enabling this option please run the playbook with -vvvv.

Thank you!

**Comment by ChristopherJHart** [comment]
@NilashishC The configuration of the switch is very vanilla, in this case. The full running-config is below:
```
N9K# show running-config 

!Command: show running-config
!Running configuration last done at: Wed Dec  2 17:06:15 2020
!Time: Thu Dec  3 16:44:45 2020

version 9.3(6) Bios:version 07.68 
hostname N9K
vdc N9K id 1
  limit-resource vlan minimum 16 maximum 4094
  limit-resource vrf minimum 2 maximum 4096
  limit-resource port-channel minimum 0 maximum 511
  limit-resource u4route-mem minimum 248 maximum 248
  limit-resource u6route-mem minimum 96 maximum 96
  limit-resource m4route-mem minimum 58 maximum 58
  limit-resource m6route-mem minimum 8 maximum 8

feature ospf

username admin password 5 $5$JNHLAK$JjWhWWLfUPsTXhbGFSwCzZFxtdR5d6NJ5frJF7Ij9Y/  role network-admin
username ansible password 5 $5$EGENOH$OicafHZ6mADM9Adv28mnGK1rKlt6JAi8/Xfvxa4JHU4  role network-admin
username ansible passphrase  lifetime 99999 warntime 14 gracetime 3
ip domain-lookup
copp profile strict
snmp-server user admin network-admin auth md5 0xf7ab2859012cfb7cfc33ed9c86e3b1dc priv 0xf7ab2859012cfb7cfc33ed9c86e3b1dc localizedkey
snmp-server user ansible network-admin auth md5 0xaf8334cbb60abd486018e92362624de7 priv 0xaf8334cbb60abd486018e92362624de7 localizedkey
rmon event 1 log trap public description FATAL(1) owner PMON@FATAL
rmon event 2 log trap public description CRITICAL(2) owner PMON@CRITICAL
rmon event 3 log trap public description ERROR(3) owner PMON@ERROR
rmon event 4 log trap public description WARNING(4) owner PMON@WARNING
rmon event 5 log trap public description INFORMATION(5) owner PMON@INFO

vlan 1

vrf context management
  ip route 0.0.0.0/0 10.122.140.65
hardware profile portmode 48x25G+2x100G+4x40G



interface Ethernet1/1
  ip address 192.168.1.1/24
  no shutdown

interface Ethernet1/2

interface Ethernet1/3

interface Ethernet1/4

interface Ethernet1/5

interface Ethernet1/6

interface Ethernet1/7

interface Ethernet1/8

interface Ethernet1/9

interface Ethernet1/10

interface Ethernet1/11

interface Ethernet1/12

interface Ethernet1/13

interface Ethernet1/14

interface Ethernet1/15

interface Ethernet1/16

interface Ethernet1/17

interface Ethernet1/18

interface Ethernet1/19

interface Ethernet1/20

interface Ethernet1/21

interface Ethernet1/22

interface Ethernet1/23

interface Ethernet1/24

interface Ethernet1/25

interface Ethernet1/26

interface Ethernet1/27

interface Ethernet1/28

interface Ethernet1/29

interface Ethernet1/30

interface Ethernet1/31

interface Ethernet1/32

interface Ethernet1/33

interface Ethernet1/34

interface Ethernet1/35

interface Ethernet1/36

interface Ethernet1/37

interface Ethernet1/38

interface Ethernet1/39

interface Ethernet1/40

interface Ethernet1/41

interface Ethernet1/42

interface Ethernet1/43

interface Ethernet1/44

interface Ethernet1/45

interface Ethernet1/46

interface Ethernet1/47

interface Ethernet1/48

interface Ethernet1/49

interface Ethernet1/50

interface Ethernet1/51

interface Ethernet1/52

interface Ethernet1/53

interface Ethernet1/54

interface mgmt0
  vrf member management
  ip address 10.122.140.125/26
icam monitor scale

line console
line vty
boot nxos bootflash:/nxos.9.3.6.bin 
router ospf UNDERLAY
```
The verbose output with persistent logging enabled can be found here: https://gist.github.com/ChristopherJHart/f548d49abd81ec81066403d542ab21b0

Thank you!

**Comment by ChristopherJHart** [comment]
As a side note, this issue also appears to impact loopback interfaces as well.

**Comment by ubajze** [comment]
I found out that this works only when you use `secondaries: True` under `area` configuration.

**Comment by NilashishC** [comment]
@ChristopherJHart The gist you shared only has the output of playbook run with -vvvv. The actual persistent logs messages is stored in the file that's set by ANSIBLE_LOG_PATH, as mentioned in the guide. 

However, I set up a N9Kv running 9.3.6 and tried executing the same playbook with vanilla on-box config and it still worked fine for me. Debugging this might need placing debug statements in multiple places in the code and inspecting the results to see what's exactly happening. But the persistent logs should give some more insights. Have you tried running the playbook against a different appliance and see if it works there?

**Comment by NilashishC** [comment]
@ubajze That shouldn't be the case. Can you share the playbook that you're running?

**Comment by ubajze** [comment]
@NilashishC :

Here is my OSPF config:

```
nxos# show running-config ospf
                                ^
% Invalid command at '^' marker.
```

So, as you see there is no OSPF config.

Here are my tasks:

```
    - name: "Enable OSPF feature"
      cisco.nxos.nxos_feature:
        feature: ospf
        state: enabled

    - name: "Configure OSPF process"
      cisco.nxos.nxos_ospfv2:
        config:
          processes:
            - process_id: "{{ ospf.process_id }}"

    - name: "Configure OSPF interfaces"
      cisco.nxos.nxos_ospf_interfaces:
        config:
          - name: "{{ item.name }}"
            address_family:
              - afi: "ipv4"
                processes:
                  - process_id: "{{ ospf.process_id }}"
                    area:
                      area_id: "0.0.0.0"
                      #secondaries: no
      loop: "{{ interfaces }}"
      when: item.ospf == 'enabled'
```

Notice that `secondaries`  is commented.

If I run my playbook:

``` 
TASK [Configure OSPF interfaces] ***********************************************
ok: [nxos] => (item={'name': 'Loopback0', 'ip_address': 192.168.1.1/32', 'ospf': 'enabled'})
ok: [nxos] => (item={'name': 'Ethernet1/1', 'ip_address': '192.168.100.1/30', 'ospf': 'enabled'})
```

Config:
```
nxos# show running-config ospf

feature ospf

router ospf 1
```

As you can see, no OSPF on interfaces.

Now I enable `secondaries`:

```
    - name: "Configure OSPF interfaces"
      cisco.nxos.nxos_ospf_interfaces:
        config:
          - name: "{{ item.name }}"
            address_family:
              - afi: "ipv4"
                processes:
                  - process_id: "{{ ospf.process_id }}"
                    area:
                      area_id: "0.0.0.0"
                      secondaries: no
      loop: "{{ interfaces }}"
      when: item.ospf == 'enabled'
```

Re-run:

```
TASK [Configure OSPF interfaces] ***********************************************
changed: [nxos] => (item={'name': 'Loopback0', 'ip_address': '192.168.1.1/32', 'ospf': 'enabled'})
changed: [nxos] => (item={'name': 'Ethernet1/1', 'ip_address': '192.168.100.1/30', 'ospf': 'enabled'})
```

OSPF is now also enabled on the interfaces:

```
nxos# show running-config ospf

feature ospf

router ospf 1

interface loopback0
  ip router ospf 1 area 0.0.0.0 secondaries none

interface Ethernet1/3
  ip router ospf 1 area 0.0.0.0 secondaries none
```

**Comment by NilashishC** [comment]
@ubajze Thanks for sharing the playbook. Even this worked fine for me. I ran it with Python 3.8.5 and multiple NX-OS versions (9.3.3, 9.3.5, 9.3.6, and 7.0(3)I7(1)) on N9Kv chassis. Like I mentioned to @ChristopherJHart, I'd need the persistent logs of your playbook run to debug this further since I'm not able to reproduce it myself.

cc @mikewiebe @praveenramoorthy - could you please have a look at this when you get a chance? I'm wondering if it's 
only happening for this particular chassis + NX-OS combination. Thanks!

**Comment by ChristopherJHart** [comment]
@NilashishC Apologies for the delay in getting back to you on this. A gist with the contents of the file set by ANSIBLE_LOG_PATH can be found here:

https://gist.github.com/ChristopherJHart/c595b977df0fc853259f55d45ab7e473

The contents of the playbook at the moment are below:
```yaml
- name: Configure OSPF
  connection: network_cli
  gather_facts: false
  hosts: all
  tasks:
    - name: Configure OSPF underlay process
      cisco.nxos.nxos_ospfv2:
        config:
          processes:
          - process_id: UNDERLAY 
    - name: Configure underlay interface
      cisco.nxos.nxos_interfaces:
        config:
        - name: "Ethernet1/1"
          mode: layer3
          enabled: true
    - name: Configure IP on underlay interface
      cisco.nxos.nxos_l3_interfaces:
        config:
        - name: "Ethernet1/1"
          ipv4:
          - address: "192.168.1.1/24"
    - name: Configure OSPF on underlay interface
      cisco.nxos.nxos_ospf_interfaces:
        config:
        - name: "Ethernet1/1"
          address_family:
          - afi: ipv4
            processes:
            - process_id: UNDERLAY
              area:
                area_id: 0.0.0.0
  vars:
    ansible_persistent_log_messages: True
```
This gist was created when running the playbook with -vvvv, as shown below:

```
chart2@chart2-ubuntu-playground:~/Documents/Ansible/ansible-ospf-bug$ ansible-playbook -i inventories/hosts playbook.yml -vvvv
```
Thank you!

**Comment by NilashishC** [comment]
@ChristopherJHart Thanks for sharing the logs. I'm working with @praveenramoorthy to further debug your environment.

**Comment by NilashishC** [comment]
@ChristopherJHart After a debugging session with @praveenramoorthy, we found out the root cause of this issue.

The commands for this module are rendered by Jinja2. In your environment, this was not working as expected even though the module was calculating the diff correctly. In spite of using a virtual environment (which has Jinja2 v2.11.2 installed), the `ansible_python_interpreter` option was not set. This resulted in the system python being used which has Jinja2 v2.10.2 installed. There is an issue in Jinja2 v2.10.x with using the shortcut version of ternary operator in templates that results in an `UndefinedError` exception (https://github.com/pallets/jinja/issues/710). This issue was resolved starting from Jinja2 v2.11.0 (https://github.com/pallets/jinja/pull/1079). The template for rendering the area command has such a shortcut ternary operator, thereby causing Jinja2 v2.10.2 to not render the command properly. We added `ansible_python_interpreter` in ansible.cfg and that fixed the problem since it was now using the correct python version which has Jinja2 v2.11.2 installed.

**Comment by ChristopherJHart** [comment]
@NilashishC @praveenramoorthy I appreciate all of your work on this! I confirmed that by setting `ansible_python_interpreter` in my virtual environment, this issue is resolved.

Thank you again for all of your hard work on this! I'm sorry that the root cause of this ended up being a niche aspect of my environment instead of something much more interesting.
